src/providers/WorkflowCore.Persistence.EntityFramework/Services/LargeDataOptimizedEntityFrameworkPersistenceProvider.cs

```diff
@@ -0,0 +1,135 @@
+using WorkflowCore.Persistence.EntityFramework.Interfaces;
+using System.Threading;
+using WorkflowCore.Interface;
+
+namespace WorkflowCore.Persistence.EntityFramework.Services
+{
+    public sealed class LargeDataOptimizedEntityFrameworkPersistenceProvider : EntityFrameworkPersistenceProvider, IPersistenceProvider
+    {
+        private readonly IWorkflowDbContextFactory _contextFactory;
+
+        public LargeDataOptimizedEntityFrameworkPersistenceProvider(IWorkflowDbContextFactory contextFactory, bool canCreateDb, bool canMigrateDb)
+            : base(contextFactory, canCreateDb, canMigrateDb)
+        {
+            _contextFactory = contextFactory;
+        }
+
+        /// <inheritdoc/>
+        public new async Task<IEnumerable<WorkflowInstance>> GetWorkflowInstances(WorkflowStatus? status, string type, DateTime? createdFrom, DateTime? createdTo, int skip, int take)
+        {
+            using (var db = _contextFactory.Build())
+            {
+                IQueryable<PersistedWorkflow> query = db.Set<PersistedWorkflow>()
+                    .Include(wf => wf.ExecutionPointers)
+                    .ThenInclude(ep => ep.ExtensionAttributes)
+                    .Include(wf => wf.ExecutionPointers)
+                    .AsSplitQuery()
+                    .AsQueryable();
+
+                if (status.HasValue)
+                {
+                    query = query.Where(x => x.Status == status.Value);
+                }
+
+                if (!string.IsNullOrEmpty(type))
+                {
+                    query = query.Where(x => x.WorkflowDefinitionId == type);
+                }
+
+                if (createdFrom.HasValue)
+                {
+                    query = query.Where(x => x.CreateTime >= createdFrom.Value);
+                }
+
+                if (createdTo.HasValue)
+                {
+                    query = query.Where(x => x.CreateTime <= createdTo.Value);
+                }
+
+                var rawResult = await query.OrderBy(x => x.PersistenceId).Skip(skip).Take(take).ToListAsync();
+
+                var result = new List<WorkflowInstance>(rawResult.Count);
+
+                foreach (var item in rawResult)
+                {
+                    result.Add(item.ToWorkflowInstance());
+                }
+
+                return result;
+            }
+        }
+
+        /// <inheritdoc/>
+        public new async Task<WorkflowInstance> GetWorkflowInstance(string id, CancellationToken cancellationToken = default)
+        {
+            using (var db = _contextFactory.Build())
+            {
+                var uid = new Guid(id);
+                var raw = await db.Set<PersistedWorkflow>()
+                    .Include(wf => wf.ExecutionPointers)
+                    .ThenInclude(ep => ep.ExtensionAttributes)
+                    .Include(wf => wf.ExecutionPointers)
+                    .AsSplitQuery()
+                    .FirstAsync(x => x.InstanceId == uid, cancellationToken);
```

Comment:
```suggestion
                    .FirstOrDefaultAsync(x => x.InstanceId == uid, cancellationToken);
```
because it seems you're checking for `raw` being not null